spark/src/main/java/org/apache/iceberg/spark/data/SparkValueReaders.java

```diff
@@ -305,7 +305,7 @@ private UnionReader(Schema schema, List<ValueReader<?>> readers) {
     }
 
     @Override
-    public InternalRow read(Decoder decoder, Object reuse) throws IOException {
+    public Object read(Decoder decoder, Object reuse) throws IOException {
```

Comment:
why is this needed if you are still returning `struct` which is a `InternalRow` ?

Reply:
See line 323, if it's null data, we will return `null` directly, so this method returns either `null` or an `InternalRow`, so the return type should only be `Object`. Please see the same method signature in `ValueReaders`.

Reply:
So why not returning null directly at line 323 if `nullIndex` is hit ?

Reply:
`readers[index].read(decoder, reuse)` will return `null`, I just feel writing this way will make the code more consistent with later part.

Reply:
Hmm not sure that's worthwhile with the cost of changing the method signature ... But I don't have strong opinion on this either.

Reply:
I think using `InternalRow` is better, let me make the change accordingly. Thanks for spotting this!